access/types/forms.py

```diff
@@ -38,6 +38,13 @@ def __init__(self, *args, **kwargs):
             self.show_correct_once = False
         self.request = kwargs.pop('request') if 'request' in kwargs else None
         kwargs['label_suffix'] = ''
+
+        # Set form-field ids based on exercise-key
+        exercise_key = self.exercise.get("key", False)
+        if not exercise_key:
+            exercise_key = self.exercise.get("title", "")
+        kwargs.setdefault("auto_id", "{}_%s".format(exercise_key))
```

Comment:
When the same questionnaire is open in the exercise frame and in a modal dialog (past submission), then these ids will be duplicated. Hence, it would be beneficial to add a bit of randomness to it. Technically ordinar number should work too, but there can be cases where that assumption breaks.

I started using format `exercise-{{ form_random_id }}-field-{{ field.name }}` for field ids and I created form html with

```django
<form id="exercise-{{ form_random_id }}-form" data-form-nonce="{{ form_random_id }}">
```

where

```python
from django.utils.crypto import get_random_string
form_random_id = get_random_string(length=8)
```

Thus, maybe?

```python
random_id = get_random_string(length=8)
kwargs.setdefault("auto_id", "exercise-{}-field-%s".format(random_id))
super().__init__(*args, **kwargs)
self.form_id = "exercise-{}-form".format(random_id)
self.form_nonce = random_id
```

and then you can use `form.form_id` and `form.form_nonce` in the html template.

Reply:
Oh, good to know. I followed the example in mooc-jutut, but did not realize the specificity of questionnaire in this case. I now implemented this version, works nicely :)